File: 0020-valid-parentheses/0020-valid-parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        stack<char> storage;
        
        for (auto c : s){
            if ((c == '(') || (c == '{') || (c == '[')){
                storage.push(c);
            } else if ((c == ')') || (c== '}') || (c == ']')){
                if (storage.empty()){return false;}
                
                char top = storage.top();
                storage.pop();
                
                if (top == '(' && c == ')'){continue;} 
                if (top == '{' && c == '}'){continue;}
                if (top == '[' && c == ']'){continue;}
                return false;
            }
        }
        
        if (storage.empty()){
            return true;
        }
        return false;
    }
};

File: 0020-valid-parentheses/0020-valid-parentheses.py
class Solution(object):
    def isValid(self, s):
        """
        :type s: str
        :rtype: bool
        """
        mapper = {"(":")", "{":"}", "[":"]"}
        queue = []
        for c in s:
            if c in mapper:
                queue.append(c)
            else:
                if not queue:
                    return False
                top = queue.pop()
                if mapper[top] != c:
                    return False
        if queue:
            return False
        return True


        